tests/PathBrokerTest.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+namespace Northwoods\Broker;
+
+use League\Container\Container;
+use PHPUnit\Framework\TestCase;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Zend\Diactoros\Response\HtmlResponse;
+use Zend\Diactoros\ServerRequest;
+
+class PathBrokerTest extends TestCase
+{
+    use CanMock;
+
+    /**
+     * @var Broker
+     */
+    private $broker;
+
+    public function setUp()
+    {
+        $this->broker = new Broker($this->makeContainer());
+    }
+
+    public function testRoutingToPathsWorks()
+    {
+        // Execute
+        $this->broker->paths([
+            '/foo' => new stubs\ReturnsGivenString('something'),
+            '/bar' => stubs\ReturnsGivenString::class,
```

Comment:
Please use [phony](http://eloquent-software.com/phony/latest/#mocks) for mocking instead of creating stubs.

Reply:
But I am not mocking anything. (I don't care whether and how the things are called, I just check for the expected results.)

Reply:
It's not necessary to use Phony call checks, I just don't want stubs created when a mock implementation works equally well without adding more files.

Reply:
Okay, it's your call. Will change this soon-ish.

Reply:
Hmm. I find it super painful having to mock a middleware that returns a mocked response with a mocked body. Here is what I came up with... Am I missing something?

```php
    /**
     * @return \Eloquent\Phony\Mock\Handle\InstanceHandle
     */
    private function mockMiddlewareReturning(string $responseBody)
    {
        $stream = Phony::mock(StreamInterface::class);
        $stream->__toString->returns($responseBody);
        
        $response = Phony::mock(ResponseInterface::class);
        $response->getBody->returns($stream);
        
        $middleware = Phony::mock(MiddlewareInterface::class);
        $middleware->process->returns($response);

        return $middleware;
    }
```

This is why I like stubs: they are real implementations, and usually super straight-forward. 

If the only downside is having one more file in the test code, is that really a downside? (The stub is small enough to be put into the same file, if you prefer that...)

Reply:
Always an option to use an [anonymous class](https://secure.php.net/manual/en/language.oop5.anonymous.php) if the mocking is too much work. Which I see is already being done in some places.

I would also argue that for many of these tests using Phony mocks and call checks would provide better overall testing because the expectations inside the middleware should remain consistent. For instance in `PathPrefixingHandler` the expectation is that `UriInterface` is going to be manipulated. The test should reflect this, rather than testing the output the middleware.

Reply:
But I don't want to test *how* the `PathDispatcher` does its work, I just wan't to test *that* it does it. That means two things:

1.) Ensure it dispatches to the correct middleware. I do this by comparing the expected responses, depending on which request I pipe through the broker.
2.) Ensure that it provides a manipulated request object to the middleware. I do this by returning the path (which is the part that should be manipulated) in the middleware response so that I can verify on the "outside" (a.k.a. from the result of the operation).

Mocking and call checks would couple me to a specific implementation, as far as I can tell, and that's what I wanted to avoid. Can you provide an example of how you would mock this without these downsides?